File: stacks_and_queues/1-queue.js
/*
A queue uses FIFO (first-in first-out) ordering.

Operations:

- add(item) Add an item to the end of the queue
- remove() Remove the first item from the queue
- peek() Return the top of the queue
- isEmpty() Return true if the queue is empty
*/

class Queue extends Array {

    add (item) {
        this.push(item)
    }

    remove () {
        // Slow because all of the other elements have to be shifted by one.
        this.shift()
    }

    peek () {
        let top = this[this.length - 1]
        console.log(top)
        return top
    }

    isEmpty () {
        return this.length === 0
    }

}

let q = new Queue()

console.log(q.isEmpty())

q.add(2)
q.add(5)
q.peek()

console.log(q.isEmpty())

console.log(q.length)

File: stacks_and_queues/1-queue.py
"""
https://docs.python.org/3/tutorial/datastructures.html#using-lists-as-queues
"""

class Queue:

    def __init__(self):
        self.storage = []

    def add(self, val):
        self.storage.append(val)

    def remove(self):
        # Slow because all of the other elements have to be shifted by one.
        self.storage.pop(0)

    def peak(self):
        return None if self.is_empty() else self.storage[-1]

    def is_empty(self):
        return len(self.storage) == 0

    def __str__(self):
        return str(self.storage)

q = Queue()

q.add('1')
q.add('2')
q.add('3')

print(q.peak())

print(q)
q.remove()
print(q)
q.remove()
print(q)
q.remove()
print(q)
